File: clients/client-cli/cmd/autopilot_test.go
package cmd

import (
	"strings"
	"testing"
)

// Pop autopilot, queenctl side.
//
// The SDK's one rule is that an explicit user value is sacred, and a CLI is
// where that rule is easiest to break: a flag has a DEFAULT, and a default is
// not a user value. Before autopilot the two were indistinguishable on the wire
// — an omitted `partitions` meant 1, which is exactly what the flag's default
// said — so `if maxParts > 1` was a correct gate. It stopped being one the
// moment an omitted `partitions` started meaning "broker, you choose": that
// gate would have handed a typed `--max-partitions 1` to the autopilot to
// widen, silently turning a single-partition claim into a sweep.
//
// So the gate is now Flags().Changed, and these tests pin both halves of it
// against the real cobra tree and the fake broker in conflation_test.go.

// --- --max-partitions: typed pins, untouched delegates ----------------------

func TestPopPinsTypedMaxPartitionsEvenAtOne(t *testing.T) {
	// `pop` pins the batch too (--limit defaults to 1), so with the sweep width
	// typed as well there is nothing left for autopilot to decide: the request
	// is the byte-identical legacy one, no autopilot flag, and the broker
	// applies its own default of one partition. THAT is the pin — a request
	// that never delegated anything cannot be widened.
	fb := newFakeBroker(t, popBody("ap-pop-pin", ""))
	stop := captureStdio(t)

	err := runCLI(t, fb.url(), "pop", "ap-pop-pin", "--cg", "workers",
		"--max-partitions", "1", "--wait=false", "--timeout", "2s")
	stop()
	if err != nil {
		t.Fatalf("pop: %v", err)
	}

	q := fb.firstPop(t).Query
	if _, present := q["autopilot"]; present {
		t.Fatalf("autopilot must not be sent when both knobs are pinned (raw: %v)", q.Encode())
	}
	if got, present := q["partitions"], q.Has("partitions"); present && got[0] != "1" {
		t.Fatalf("partitions = %q, want absent or \"1\" (raw: %v)", got[0], q.Encode())
	}
}

func TestPopSendsTypedMaxPartitionsOfOneWhenTheBatchIsDelegated(t *testing.T) {
	// Same typed --max-partitions 1, but with --limit 0 the batch is nobody's
	// decision, so autopilot engages for it -- and now the pin HAS to travel,
	// because an omitted partitions would be the broker's to choose.
	fb := newFakeBroker(t, popBody("ap-pop-pin-explicit", ""))
	stop := captureStdio(t)

	err := runCLI(t, fb.url(), "pop", "ap-pop-pin-explicit", "--cg", "workers",
		"--max-partitions", "1", "--limit", "0", "--wait=false", "--timeout", "2s")
	stop()
	if err != nil {
		t.Fatalf("pop: %v", err)
	}

	q := fb.firstPop(t).Query
	if got := q.Get("partitions"); got != "1" {
		t.Fatalf("partitions = %q, want \"1\": a typed --max-partitions 1 must pin the claim (raw: %v)", got, q.Encode())
	}
	if got := q.Get("autopilot"); got != "true" {
		t.Fatalf("autopilot = %q, want \"true\" (raw: %v)", got, q.Encode())
	}
	if _, present := q["batch"]; present {
		t.Fatalf("batch must not be sent when --limit is 0 and --batch untyped (raw: %v)", q.Encode())
	}
}

func TestPopLeavesMaxPartitionsToTheBrokerWhenUntouched(t *testing.T) {
	fb := newFakeBroker(t, popBody("ap-pop-auto", ""))
	stop := captureStdio(t)

	err := runCLI(t, fb.url(), "pop", "ap-pop-auto", "--cg", "workers",
		"--wait=false", "--timeout", "2s")
	stop()
	if err != nil {
		t.Fatalf("pop: %v", err)
	}

	q := fb.firstPop(t).Query
	if _, present := q["partitions"]; present {
		t.Fatalf("partitions must not be sent when the flag was not typed (raw: %v)", q.Encode())
	}
	if got := q.Get("autopilot"); got != "true" {
		t.Fatalf("autopilot = %q, want \"true\" (raw: %v)", got, q.Encode())
	}
	// pop's --limit defaults to 1 and always pins the batch: that is this
	// command's own semantics and autopilot does not touch it.
	if got := q.Get("batch"); got != "1" {
		t.Fatalf("batch = %q, want \"1\" (--limit still pins it) (raw: %v)", got, q.Encode())
	}
}

func TestPopPassesTypedMaxPartitionsAboveOne(t *testing.T) {
	fb := newFakeBroker(t, popBody("ap-pop-wide", ""))
	stop := captureStdio(t)

	err := runCLI(t, fb.url(), "pop", "ap-pop-wide", "--cg", "workers",
		"--max-partitions", "16", "--wait=false", "--timeout", "2s")
	stop()
	if err != nil {
		t.Fatalf("pop: %v", err)
	}

	q := fb.firstPop(t).Query
	if got := q.Get("partitions"); got != "16" {
		t.Fatalf("partitions = %q, want \"16\" (raw: %v)", got, q.Encode())
	}
	// Both knobs pinned (batch via --limit) leaves autopilot nothing to do, so
	// the request is the one queenctl sent before this feature existed.
	if _, present := q["autopilot"]; present {
		t.Fatalf("autopilot must not be sent when both knobs are pinned (raw: %v)", q.Encode())
	}
}

func TestTailPinsTypedMaxPartitionsEvenAtOne(t *testing.T) {
	// `tail` runs the consume loop, a DIFFERENT param builder from the one
	// `pop` uses. A pop-only test would pass with a tail that dropped the pin.
	fb := newFakeBroker(t, popBody("ap-tail-pin", ""))
	stop := captureStdio(t)

	err := runCLI(t, fb.url(), "tail", "ap-tail-pin", "--cg", "workers",
		"--max-partitions", "1", "-n", "1", "--timeout", "2s")
	stdout, _ := stop()
	if err != nil {
		t.Fatalf("tail: %v", err)
	}
	if !strings.Contains(stdout, `"transactionId"`) {
		t.Fatalf("tail printed no message; stdout=%q", stdout)
	}

	q := fb.firstPop(t).Query
	if got := q.Get("partitions"); got != "1" {
		t.Fatalf("partitions = %q, want \"1\": a typed --max-partitions 1 must pin the claim (raw: %v)", got, q.Encode())
	}
}

func TestTailLeavesBothKnobsToTheBrokerWhenUntouched(t *testing.T) {
	// tail's --batch defaults to 0 (unset) and --max-partitions is untyped, so
	// this is the fully delegated shape: autopilot=true and neither knob.
	fb := newFakeBroker(t, popBody("ap-tail-auto", ""))
	stop := captureStdio(t)

	err := runCLI(t, fb.url(), "tail", "ap-tail-auto", "--cg", "workers",
		"-n", "1", "--timeout", "2s")
	stop()
	if err != nil {
		t.Fatalf("tail: %v", err)
	}

	q := fb.firstPop(t).Query
	if got := q.Get("autopilot"); got != "true" {
		t.Fatalf("autopilot = %q, want \"true\" (raw: %v)", got, q.Encode())
	}
	for _, absent := range []string{"partitions", "batch"} {
		if _, present := q[absent]; present {
			t.Fatalf("%s must not be sent when the flag was not typed (raw: %v)", absent, q.Encode())
		}
	}
}

func TestTailPinsTypedBatch(t *testing.T) {
	fb := newFakeBroker(t, popBody("ap-tail-batch", ""))
	stop := captureStdio(t)

	err := runCLI(t, fb.url(), "tail", "ap-tail-batch", "--cg", "workers",
		"--batch", "25", "-n", "1", "--timeout", "2s")
	stop()
	if err != nil {
		t.Fatalf("tail: %v", err)
	}

	q := fb.firstPop(t).Query
	if got := q.Get("batch"); got != "25" {
		t.Fatalf("batch = %q, want \"25\" (raw: %v)", got, q.Encode())
	}
	// Batch pinned, sweep width still the broker's.
	if got := q.Get("autopilot"); got != "true" {
		t.Fatalf("autopilot = %q, want \"true\" (raw: %v)", got, q.Encode())
	}
	if _, present := q["partitions"]; present {
		t.Fatalf("partitions must not be sent when the flag was not typed (raw: %v)", q.Encode())
	}
}

// --- the harness itself -----------------------------------------------------

func TestTypedFlagDoesNotLeakIntoTheNextRun(t *testing.T) {
	// pflag records "was it typed" on the flag object, and the cobra tree is a
	// package singleton: without the reset in resetConflationFlags, the run
	// below would inherit the previous run's --max-partitions and pin a claim
	// nobody asked to pin. This test exists because that failure is invisible —
	// the second run simply sends a parameter, and every assertion about it
	// passes for the wrong reason.
	fb := newFakeBroker(t, popBody("ap-leak-1", ""), popBody("ap-leak-2", ""))
	stop := captureStdio(t)

	if err := runCLI(t, fb.url(), "pop", "ap-leak-1", "--cg", "workers",
		"--max-partitions", "8", "--wait=false", "--timeout", "2s"); err != nil {
		t.Fatalf("first pop: %v", err)
	}
	if err := runCLI(t, fb.url(), "pop", "ap-leak-2", "--cg", "workers",
		"--wait=false", "--timeout", "2s"); err != nil {
		t.Fatalf("second pop: %v", err)
	}
	stop()

	pops := fb.pops()
	if len(pops) != 2 {
		t.Fatalf("expected 2 pops, got %d", len(pops))
	}
	if got := pops[0].Query.Get("partitions"); got != "8" {
		t.Fatalf("first pop partitions = %q, want \"8\"", got)
	}
	if _, present := pops[1].Query["partitions"]; present {
		t.Fatalf("the typed flag leaked into the next run (raw: %v)", pops[1].Query.Encode())
	}
}

File: clients/client-php/tests/AutopilotTest.php
<?php

namespace Queen\Tests;

use GuzzleHttp\HandlerStack;
use PHPUnit\Framework\TestCase;
use Psr\Http\Message\RequestInterface;
use Queen\Queen;
use Queen\Support\Defaults;
use Queen\Support\PopAutopilot;
use Queen\Tests\Support\PlanHandler;

/**
 * Pop autopilot, the PHP row — the broker sizes the knobs the caller did not.
 *
 * The four things a client can be wrong about here, and why each is asserted
 * against the WHOLE query string rather than against one parameter:
 *
 *  1. EVERY BUILDER MUST AGREE. PHP has THREE pop param builders — the inline
 *     one in QueueBuilder::pop, ConsumerManager::buildParams, and a
 *     byte-identical copy in HighLevelConsumer::buildParams — which is the
 *     standing hazard in this family of clients and the reason the emission rule
 *     lives in one class. Every case below is run through all three, and all
 *     three are compared to the same expected string.
 *  2. NOT ENGAGING AUTOPILOT MUST BE BYTE-IDENTICAL TO THE OLD SDK. The escape
 *     hatch is only worth having if it is exact, and "exact" is not something a
 *     test of one parameter can show: a stray autopilot=true, or a batch that
 *     stopped being emitted, is a different request. Hence full-string equality
 *     including the parameters this feature never touches, and including their
 *     ORDER — this SDK builds an array and http_build_query's it, so order is
 *     part of the bytes.
 *  3. AN EXPLICIT VALUE IS SACRED, PER DIMENSION. partitions(1) and "never
 *     called partitions" both used to reach the wire as nothing at all; they are
 *     now different requests, and the pinned one must survive autopilot.
 *  4. THE ADDITIVE RESPONSE FIELD MUST NOT BE LOAD-BEARING. A broker that does
 *     not send it, sends it half-filled, or sends it with fields this SDK has
 *     never heard of, all have to work.
 */
class AutopilotTest extends TestCase
{
    /**
     * The shared spine of every case: a named queue and group, no long poll,
     * default timeout. Everything that varies below is sizing.
     */
    private const TAIL = 'wait=false&timeout=30000&consumerGroup=workers';

    protected function tearDown(): void
    {
        putenv(PopAutopilot::ENV_VAR);
        unset($_ENV[PopAutopilot::ENV_VAR], $_SERVER[PopAutopilot::ENV_VAR]);
    }

    /**
     * @return array<string, array{\Closure(\Queen\Builders\QueueBuilder): \Queen\Builders\QueueBuilder, string}>
     */
    public static function sizingCases(): array
    {
        $tail = self::TAIL;

        return [
            // Nothing set: both knobs go to the broker, neither travels.
            'nothing set' => [fn($b) => $b, "autopilot=true&{$tail}"],
            // A pinned width travels, the batch is delegated.
            'partitions only' => [fn($b) => $b->partitions(4), "autopilot=true&{$tail}&partitions=4"],
            // The pin that used to be indistinguishable from unset. partitions(1)
            // is a decision — hold this consumer to one partition — and the
            // broker has to be told, or autopilot would widen it.
            'partitions pinned to one' => [fn($b) => $b->partitions(1), "autopilot=true&{$tail}&partitions=1"],
            // A pinned batch travels, the sweep width is delegated.
            'batch only' => [fn($b) => $b->batch(50), "autopilot=true&batch=50&{$tail}"],
            // Both set: nothing left to decide, so no autopilot parameter and the
            // exact request the pre-autopilot SDK sent.
            'both set' => [fn($b) => $b->batch(50)->partitions(4), "batch=50&{$tail}&partitions=4"],
            // ...including the one where the old SDK never emitted partitions=1.
            'both set, partitions one' => [fn($b) => $b->batch(50)->partitions(1), "batch=50&{$tail}"],
            // The escape hatch: the client-side defaults are back.
            'autopilot off, nothing set' => [fn($b) => $b->autopilot(false), "batch=1&{$tail}"],
            'autopilot off, partitions one' => [fn($b) => $b->autopilot(false)->partitions(1), "batch=1&{$tail}"],
            'autopilot off, both set' => [
                fn($b) => $b->autopilot(false)->batch(50)->partitions(4),
                "batch=50&{$tail}&partitions=4",
            ],
            // autopilot(true) is the default spelled out: it must change nothing.
            'autopilot on, both set' => [
                fn($b) => $b->autopilot(true)->batch(50)->partitions(4),
                "batch=50&{$tail}&partitions=4",
            ],
            // batch(0) is not "a batch of zero" and never was: it is the absence
            // of an opinion, which now means the broker decides.
            'batch zero is unset' => [fn($b) => $b->batch(0), "autopilot=true&{$tail}"],
        ];
    }

    // ===========================
    // 1. Param assembly, from all three builders
    // ===========================

    /** Builder #1: the inline param block in QueueBuilder::pop. */
    #[\PHPUnit\Framework\Attributes\DataProvider('sizingCases')]
    public function testPopParamAssembly(\Closure $build, string $want): void
    {
        $handler = new PlanHandler([], self::ok(self::popBody()));
        $queen = $this->queen($handler);

        $build($queen->queue('orders')->group('workers')->wait(false))->pop();

        $this->assertSame($want, self::rawQuery($handler->requests[0]));
    }

    /** Builder #2: ConsumerManager::buildParams, the callback consume loop. */
    #[\PHPUnit\Framework\Attributes\DataProvider('sizingCases')]
    public function testConsumeParamAssembly(\Closure $build, string $want): void
    {
        $handler = new PlanHandler([], self::ok(self::popBody()));
        $queen = $this->queen($handler);

        $handled = 0;
        $build($queen->queue('orders')->group('workers')->wait(false))
            ->autoAck(false)->limit(1)
            ->consume(function (array $messages) use (&$handled): void {
                $handled += count($messages);
            })
            ->execute();

        $this->assertSame(1, $handled);
        $this->assertSame($want, self::rawQuery($handler->requests[0]));
    }

    /**
     * Builder #3: HighLevelConsumer::buildParams, the rdkafka-style surface.
     *
     * It gets its own cases because it pins the BATCH itself, per call, and
     * always did: consume() hands back exactly one message, so claiming more
     * would leave the rest leased and undelivered, and consumeBatch's
     * $maxMessages IS the budget being asked for. So on this surface the batch
     * is never the broker's; the sweep width still is, unless pinned.
     *
     * That is also why this consumer rebuilds its query per call instead of
     * patching `batch` into a rendered string, which is what it used to do:
     * a patched string would have carried `autopilot=true` next to two pinned
     * knobs — a request asking the broker to decide nothing.
     */
    public function testHighLevelConsumerPinsTheBatchAndDelegatesOnlyTheWidth(): void
    {
        $handler = new PlanHandler([], self::ok(self::popBody()));
        $queen = $this->queen($handler);

        $consumer = $queen->queue('orders')->group('workers')->getConsumer();
        $consumer->subscribe();
        $consumer->consume(50);

        $this->assertSame(
            'autopilot=true&batch=1&wait=true&timeout=50&consumerGroup=workers',
            self::rawQuery($handler->requests[0])
        );
    }

    public function testHighLevelConsumerWithBothKnobsPinnedSendsNoFlag(): void
    {
        $handler = new PlanHandler([], self::ok(self::popBody()));
        $queen = $this->queen($handler);

        $consumer = $queen->queue('orders')->group('workers')->partitions(4)->getConsumer();
        $consumer->subscribe();
        $consumer->consume(50);

        $this->assertSame(
            'batch=1&wait=true&timeout=50&consumerGroup=workers&partitions=4',
            self::rawQuery($handler->requests[0]),
            'the surface pinned the batch and the caller pinned the width: nothing left to decide'
        );
    }

    public function testHighLevelConsumerBatchPinsMaxMessages(): void
    {
        $handler = new PlanHandler([], self::ok(self::popBody(messages: 2)));
        $queen = $this->queen($handler);

        $consumer = $queen->queue('orders')->group('workers')->getConsumer();
        $consumer->subscribe();
        $consumer->consumeBatch(50, 10);

        $this->assertSame(
            'autopilot=true&batch=10&wait=true&timeout=50&consumerGroup=workers',
            self::rawQuery($handler->requests[0])
        );
    }

    public function testHighLevelConsumerOffIsTheOldRequestByteForByte(): void
    {
        $handler = new PlanHandler([], self::ok(self::popBody()));
        $queen = $this->queen($handler);

        $consumer = $queen->queue('orders')->group('workers')->autopilot(false)->getConsumer();
        $consumer->subscribe();
        $consumer->consume(50);

        $this->assertSame(
            'batch=1&wait=true&timeout=50&consumerGroup=workers',
            self::rawQuery($handler->requests[0])
        );
    }

    // ===========================
    // 2. The process-wide rollback
    // ===========================

    public function testEnvVarDisablesAutopilot(): void
    {
        putenv(PopAutopilot::ENV_VAR . '=off');
        $handler = new PlanHandler([], self::ok(self::popBody()));
        $queen = $this->queen($handler);

        $queen->queue('orders')->group('workers')->wait(false)->pop();

        $this->assertSame('batch=1&' . self::TAIL, self::rawQuery($handler->requests[0]));
    }

    public function testEnvVarIsReadOnceAtConstruction(): void
    {
        $handler = new PlanHandler([], self::ok(self::popBody()));
        $queen = $this->queen($handler);

        // Set AFTER the client exists: this is a deployment-level rollback, not
        // a per-request switch, so a live client must not move.
        putenv(PopAutopilot::ENV_VAR . '=off');
        $queen->queue('orders')->group('workers')->wait(false)->pop();

        $this->assertSame('autopilot=true&' . self::TAIL, self::rawQuery($handler->requests[0]));
    }

    public function testExplicitAutopilotOutranksTheEnvironment(): void
    {
        putenv(PopAutopilot::ENV_VAR . '=off');
        $handler = new PlanHandler([], self::ok(self::popBody()));
        $queen = $this->queen($handler);

        $queen->queue('orders')->group('workers')->wait(false)->autopilot(true)->pop();

        $this->assertSame('autopilot=true&' . self::TAIL, self::rawQuery($handler->requests[0]));
    }

    public function testEnvVarVocabulary(): void
    {
        foreach (['off', 'OFF', ' off ', 'false', '0', 'no', 'disabled'] as $value) {
            putenv(PopAutopilot::ENV_VAR . '=' . $value);
            $this->assertTrue(PopAutopilot::disabledByEnv(), "{$value} should disable autopilot");
        }
        foreach (['', 'on', 'true', '1', 'yes', 'nonsense'] as $value) {
            putenv(PopAutopilot::ENV_VAR . '=' . $value);
            $this->assertFalse(PopAutopilot::disabledByEnv(), "{$value} should leave autopilot on");
        }
        putenv(PopAutopilot::ENV_VAR);
        $this->assertFalse(PopAutopilot::disabledByEnv(), 'unset leaves autopilot on');
    }

    /** php-fpm pools and Laravel's .env land in the superglobals, not in getenv(). */
    public function testEnvVarIsAlsoReadFromTheSuperglobals(): void
    {
        $_ENV[PopAutopilot::ENV_VAR] = 'off';
        $this->assertTrue(PopAutopilot::disabledByEnv());
        unset($_ENV[PopAutopilot::ENV_VAR]);

        $_SERVER[PopAutopilot::ENV_VAR] = 'off';
        $this->assertTrue(PopAutopilot::disabledByEnv());
        unset($_SERVER[PopAutopilot::ENV_VAR]);
    }

    // ===========================
    // 3. The additive response field
    // ===========================

    public function testDecisionParsesWhatTheBrokerChose(): void
    {
        $this->assertNull(PopAutopilot::decision(null));
        $this->assertNull(PopAutopilot::decision(['messages' => []]), 'absent');
        $this->assertNull(PopAutopilot::decision(['autopilot' => null]), 'null');
        $this->assertNull(PopAutopilot::decision(['autopilot' => true]), 'not an object');

        $this->assertSame(
            ['partitions' => 8, 'batch' => 200, 'waitMillis' => 25],
            PopAutopilot::decision(['autopilot' => ['partitions' => 8, 'batch' => 200, 'waitMs' => 25]])
        );

        // waitMs is optional: the broker sends it only when it has an opinion.
        $this->assertSame(
            ['partitions' => 4, 'batch' => 64, 'waitMillis' => 0],
            PopAutopilot::decision(['autopilot' => ['partitions' => 4, 'batch' => 64]])
        );

        // Forward compatibility: a newer broker growing a field must not cost
        // this client the fields it does understand.
        $this->assertSame(
            ['partitions' => 2, 'batch' => 10, 'waitMillis' => 5],
            PopAutopilot::decision([
                'autopilot' => ['partitions' => 2, 'batch' => 10, 'waitMs' => 5, 'reason' => 'ready_age'],
            ])
        );

        // A field of the wrong type is dropped, not fatal.
        $this->assertSame(
            ['partitions' => 0, 'batch' => 10, 'waitMillis' => 0],
            PopAutopilot::decision(['autopilot' => ['partitions' => 'eight', 'batch' => 10]])
        );
    }

    public function testPopResultReportsWhatTheBrokerChose(): void
    {
        $body = self::popBody();
        $body['autopilot'] = ['partitions' => 8, 'batch' => 200, 'waitMs' => 25];
        $handler = new PlanHandler([], self::ok($body));
        $queen = $this->queen($handler);

        $res = $queen->queue('orders')->group('workers')->wait(false)->popResult();

        $this->assertCount(1, $res['messages']);
        $this->assertSame(['partitions' => 8, 'batch' => 200, 'waitMillis' => 25], $res['autopilot']);
    }

    public function testPopResultIsNullWhenTheBrokerSaidNothing(): void
    {
        $handler = new PlanHandler([], self::ok(self::popBody()));
        $queen = $this->queen($handler);

        $res = $queen->queue('orders')->group('workers')->wait(false)->popResult();

        $this->assertCount(1, $res['messages']);
        $this->assertNull($res['autopilot'], 'a broker older than 1.2 is not an error');
    }

    public function testPopStillReturnsABareMessageList(): void
    {
        $body = self::popBody();
        $body['autopilot'] = ['partitions' => 8, 'batch' => 200];
        $handler = new PlanHandler([], self::ok($body));
        $queen = $this->queen($handler);

        $messages = $queen->queue('orders')->group('workers')->wait(false)->pop();

        $this->assertCount(1, $messages);
        $this->assertArrayHasKey('transactionId', reset($messages));
    }

    // ===========================
    // 4. Empty-poll pacing
    // ===========================

    public function testEmptyPollDelayHonoursTheAdviceAndFallsBack(): void
    {
        $this->assertSame(
            PopAutopilot::EMPTY_POLL_BACKOFF_MICROS,
            PopAutopilot::emptyPollDelayMicros(null)
        );
        $this->assertSame(
            PopAutopilot::EMPTY_POLL_BACKOFF_MICROS,
            PopAutopilot::emptyPollDelayMicros(['partitions' => 1, 'batch' => 1, 'waitMillis' => 0])
        );
        $this->assertSame(
            250_000,
            PopAutopilot::emptyPollDelayMicros(['partitions' => 1, 'batch' => 1, 'waitMillis' => 250])
        );
    }

    // ===========================
    // 5. The rule itself, in isolation
    // ===========================

    public function testSizingLeavesAPinnedDimensionAlone(): void
    {
        $this->assertSame(
            ['autopilot' => true, 'batch' => null, 'partitions' => null],
            PopAutopilot::sizing(null, null, 1, true)
        );
        $this->assertSame(
            ['autopilot' => true, 'batch' => '50', 'partitions' => null],
            PopAutopilot::sizing(50, null, 1, true)
        );
        $this->assertSame(
            ['autopilot' => true, 'batch' => null, 'partitions' => '1'],
            PopAutopilot::sizing(null, 1, 1, true)
        );
        // Both set: nothing to decide, so the flag does not travel either.
        $this->assertSame(
            ['autopilot' => false, 'batch' => '50', 'partitions' => '4'],
            PopAutopilot::sizing(50, 4, 1, true)
        );
        // Off: the client-side default comes back and partitions keeps its >1 gate.
        $this->assertSame(
            ['autopilot' => false, 'batch' => '1', 'partitions' => null],
            PopAutopilot::sizing(null, null, 1, false)
        );
        $this->assertSame(
            ['autopilot' => false, 'batch' => '1', 'partitions' => null],
            PopAutopilot::sizing(null, 1, 1, false)
        );
    }

    public function testBuilderSetterIsFluent(): void
    {
        $builder = (new Queen('http://queen.test'))->queue('orders');
        $this->assertSame($builder, $builder->autopilot(false));
    }

    /**
     * The queen:consume command: a knob the operator TYPED is a pin, a knob left
     * out is the broker's. A --batch default of 1 would have pinned a dimension
     * on every run without anyone asking, which is the feature switched off by
     * accident.
     */
    public function testConsumeCommandLeavesTheSizingKnobsUnsetByDefault(): void
    {
        $definition = (new \Queen\Laravel\Commands\ConsumeCommand())->getDefinition();

        $this->assertTrue($definition->hasOption('batch'));
        $this->assertNull($definition->getOption('batch')->getDefault());
        $this->assertTrue($definition->hasOption('partitions'));
        $this->assertNull($definition->getOption('partitions')->getDefault());
        $this->assertTrue($definition->hasOption('no-autopilot'));
    }

    public function testConsumeDefaultsStillCarryTheAutopilotOffValues(): void
    {
        // They are no longer applied when autopilot is on, but they are still the
        // numbers .autopilot(false) restores, and the table is what documents it.
        $this->assertSame(1, Defaults::CONSUME_DEFAULTS['batch']);
        $this->assertSame(1, Defaults::CONSUME_DEFAULTS['maxPartitions']);
        $this->assertSame(1, Defaults::POP_DEFAULTS['batch']);
    }

    // ===========================
    // Helpers
    // ===========================

    private function queen(PlanHandler $handler): Queen
    {
        return new Queen([
            'url' => 'http://queen.test',
            'retryAttempts' => 1,
            'handler' => HandlerStack::create($handler),
        ]);
    }

    private static function ok(array $json): array
    {
        return ['status' => 200, 'json' => $json];
    }

    private static function popBody(int $messages = 1): array
    {
        return [
            'messages' => array_map(fn(int $i): array => [
                'transactionId' => "tx-{$i}",
                'partitionId' => 'p1',
                'queue' => 'orders',
                'partition' => 'Default',
                'data' => ['n' => $i],
                'leaseId' => 'lease-1',
            ], $messages > 0 ? range(1, $messages) : []),
            'partitionsClaimed' => 1,
        ];
    }

    /** The query string exactly as it left the client, order included. */
    private static function rawQuery(RequestInterface $request): string
    {
        return $request->getUri()->getQuery();
    }
}
